File: lib/services/google.ex
defmodule Omnixent.Services.Google do

  @google_endpoint    "https://www.google.com/complete/search?q="
  @google_queryparams "&client=psy-ab&"

  @spec format_uri(string, atom, atom) :: string
  def format_uri(term, country, language) do
    @google_endpoint
      <> URI.encode(term)
      <> @google_queryparams
      <> "hl="
      <> String.downcase(Atom.to_string(country))
      <> "-"
      <> String.upcase(Atom.to_string(language))
  end

  @spec extract_body(binary) :: list(string)
  def extract_body(response) do
    "#{:erlang.binary_to_list(response)}"
      |> Omnixent.Utils.safe_json_decode
      |> Enum.at(1)
      |> Enum.map(& 
        &1 
        |> hd
        |> String.replace("<b>", "")
        |> String.replace("</b>", "")
        |> String.replace("&#39;", "'")
      )
  end

end

File: src/lib/services/google.ts
import { ServiceArgs } from '.';
import { normalizeServiceArgs } from '../utils';

const endpoint = 'https://www.google.com/complete/search';

export function formatURI(options: ServiceArgs): string {
  const { t, hl } = normalizeServiceArgs(options);
  return `${endpoint}?client=gws-wiz&q=${t}&hl=${hl}`;
}

export function extractBody(response: string) {
  const body = response.replace(/^window\.google\.ac\.h\(/, '').replace(/\)$/, '');

  return JSON.parse(body)
    ?.flat()
    ?.flat()
    ?.filter((el: any) => typeof el === 'string')
    ?.map((term: string) => term.replace(/<b>/gi, ''))
    ?.map((term: string) => term.replace(/<\/b>/gi, ''))
    ?.map((term: string) => term.replace(/&#39;/gi, "'"));
}
